src/NuGet.Core/NuGet.Protocol.Core.v3/Utility/LocalFolderUtility.cs

```diff
@@ -0,0 +1,753 @@
+        {
+            if (root == null)
+            {
+                throw new ArgumentNullException(nameof(root));
+            }
+
+            if (id == null)
+            {
+                throw new ArgumentNullException(nameof(id));
+            }
+
+            if (log == null)
+            {
+                throw new ArgumentNullException(nameof(log));
+            }
+
+            foreach (var path in GetNupkgsFromFlatFolder(root, log))
+            {
+                if (IsPossiblePackageMatch(path, id))
+                {
+                    yield return path;
+                }
+            }
+
+            yield break;
+        }
+
+        /// <summary>
+        /// Discover all nupkgs from a v3 folder.
+        /// </summary>
+        /// <param name="root">Folder root.</param>
+        public static IEnumerable<LocalPackageInfo> GetPackagesV3(string root, ILogger log)
+        {
+            if (root == null)
+            {
+                throw new ArgumentNullException(nameof(root));
+            }
+
+            if (log == null)
+            {
+                throw new ArgumentNullException(nameof(log));
+            }
+
+            // Validate teh root path
+            DirectoryInfo rootDirectoryInfo = GetAndVerifyRootDirectory(root);
+
+            if (!rootDirectoryInfo.Exists)
+            {
+                // Directory is missing
+                yield break;
+            }
+
+            // Match all nupkgs in the folder
+            foreach (var idPath in GetDirectoriesSafe(root, log))
+            {
+                foreach (var nupkg in GetPackagesV3(root, id: idPath.Name, log: log))
+                {
+                    yield return nupkg;
+                }
+            }
+
+            yield break;
+        }
+
+        /// <summary>
+        /// Discover nupkgs from a v3 local folder.
+        /// </summary>
+        /// <param name="root">Folder root.</param>
+        /// <param name="id">Package id or package id prefix.</param>
+        public static IEnumerable<LocalPackageInfo> GetPackagesV3(string root, string id, ILogger log)
+        {
+            if (root == null)
+            {
+                throw new ArgumentNullException(nameof(root));
+            }
+
+            if (id == null)
+            {
+                throw new ArgumentNullException(nameof(id));
+            }
+
+            if (log == null)
+            {
+                throw new ArgumentNullException(nameof(log));
+            }
+
+            // Check for package files one level deep.
+            DirectoryInfo rootDirectoryInfo = GetAndVerifyRootDirectory(root);
+
+            if (!rootDirectoryInfo.Exists)
+            {
+                // Directory is missing
+                yield break;
+            }
+
+            var pathResolver = new VersionFolderPathResolver(root);
+            var idRoot = Path.Combine(root, id);
+
+            foreach (var versionDir in GetDirectoriesSafe(idRoot, log))
```

Comment:
Why is there individual package file validation and error logging in this overload and not the other?